include/akushon/config/grpc/call_data.hpp

```diff
@@ -0,0 +1,75 @@
+// Copyright (c) 2023 Ichiro ITS
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#ifndef AKUSHON__CONFIG__GRPC__CALL_DATA_HPP_
+#define AKUSHON__CONFIG__GRPC__CALL_DATA_HPP_
+
+#include "akushon_interfaces/akushon.grpc.pb.h"
+#include "akushon_interfaces/akushon.pb.h"
+#include "grpc/support/log.h"
+#include "grpcpp/grpcpp.h"
+#include "akushon/config/grpc/call_data_base.hpp"
+
+namespace akushon
+{
+template <class ConfigRequest, class ConfigReply>
+class CallData : CallDataBase
+{
+public:
+  CallData(
+    akushon_interfaces::proto::Config::AsyncService * service, grpc::ServerCompletionQueue * cq,
+    const std::string& path) : status_(CallStatus::CREATE), service_(service), cq_(cq), responder_(&ctx_), path_(path)
+{
+}
```

Comment:
Let's put this function definition outside the class so it won't be considered as an inline function.
```cpp
template <class ConfigRequest, class ConfigReply>
class CallData : CallDataBase {
  CallData(
    akushon_interfaces::proto::Config::AsyncService * service, grpc::ServerCompletionQueue * cq,
    const std::string& path);
};

template <class ConfigRequest, class ConfigReply>
CallData<ConfigRequest, ConfigReply>::CallData(
    akushon_interfaces::proto::Config::AsyncService * service, grpc::ServerCompletionQueue * cq,
    const std::string& path) : status_(CallStatus::CREATE), service_(service), cq_(cq), responder_(&ctx_), path_(path) {
}
```

Reply:
if i move from header to cpp files, i need to add somedefinition mas like
`template class CallData<akushon_interfaces::proto::ConfigActions, akushon_interfaces::proto::Empty>;
template class CallData<akushon_interfaces::proto::Empty, akushon_interfaces::proto::ConfigActions>;
template class CallData<akushon_interfaces::proto::SetJointsData, akushon_interfaces::proto::Empty>;
template class CallData<
  akushon_interfaces::proto::SetTorquesData, akushon_interfaces::proto::Empty>;
template class CallData<akushon_interfaces::proto::Empty, akushon_interfaces::proto::CurrentJoints>;
template class CallData<
  akushon_interfaces::proto::ConfigRunAction, akushon_interfaces::proto::Empty>;`
 
if not, then I will get undefined reference errors like this
`
/usr/bin/ld: libakushon.so: undefined reference to `akushon::CallData<akushon_interfaces::proto::ConfigActions, akushon_interfaces::proto::Empty>::CallData(akushon_interfaces::proto::Config::WithAsyncMethod_GetConfig<akushon_interfaces::proto::Config::WithAsyncMethod_SaveConfig<akushon_interfaces::proto::Config::WithAsyncMethod_PublishSetJoints<akushon_interfaces::proto::Config::WithAsyncMethod_RunAction<akushon_interfaces::proto::Config::WithAsyncMethod_SetTorques<akushon_interfaces::proto::Config::WithAsyncMethod_SubscribeCurrentJoints<akushon_interfaces::proto::Config::Service> > > > > >*, grpc::ServerCompletionQueue*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: libakushon.so: undefined reference to `akushon::CallData<akushon_interfaces::proto::SetTorquesData, akushon_interfaces::proto::Empty>::CallData(akushon_interfaces::proto::Config::WithAsyncMethod_GetConfig<akushon_interfaces::proto::Config::WithAsyncMethod_SaveConfig<akushon_interfaces::proto::Config::WithAsyncMethod_PublishSetJoints<akushon_interfaces::proto::Config::WithAsyncMethod_RunAction<akushon_interfaces::proto::Config::WithAsyncMethod_SetTorques<akushon_interfaces::proto::Config::WithAsyncMethod_SubscribeCurrentJoints<akushon_interfaces::proto::Config::Service> > > > > >*, grpc::ServerCompletionQueue*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: libakushon.so: undefined reference to `akushon::CallData<akushon_interfaces::proto::ConfigRunAction, akushon_interfaces::proto::Empty>::CallData(akushon_interfaces::proto::Config::WithAsyncMethod_GetConfig<akushon_interfaces::proto::Config::WithAsyncMethod_SaveConfig<akushon_interfaces::proto::Config::WithAsyncMethod_PublishSetJoints<akushon_interfaces::proto::Config::WithAsyncMethod_RunAction<akushon_interfaces::proto::Config::WithAsyncMethod_SetTorques<akushon_interfaces::proto::Config::WithAsyncMethod_SubscribeCurrentJoints<akushon_interfaces::proto::Config::Service> > > > > >*, grpc::ServerCompletionQueue*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: libakushon.so: undefined reference to `akushon::CallData<akushon_interfaces::proto::Empty, akushon_interfaces::proto::CurrentJoints>::CallData(akushon_interfaces::proto::Config::WithAsyncMethod_GetConfig<akushon_interfaces::proto::Config::WithAsyncMethod_SaveConfig<akushon_interfaces::proto::Config::WithAsyncMethod_PublishSetJoints<akushon_interfaces::proto::Config::WithAsyncMethod_RunAction<akushon_interfaces::proto::Config::WithAsyncMethod_SetTorques<akushon_interfaces::proto::Config::WithAsyncMethod_SubscribeCurrentJoints<akushon_interfaces::proto::Config::Service> > > > > >*, grpc::ServerCompletionQueue*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: libakushon.so: undefined reference to `akushon::CallData<akushon_interfaces::proto::SetJointsData, akushon_interfaces::proto::Empty>::CallData(akushon_interfaces::proto::Config::WithAsyncMethod_GetConfig<akushon_interfaces::proto::Config::WithAsyncMethod_SaveConfig<akushon_interfaces::proto::Config::WithAsyncMethod_PublishSetJoints<akushon_interfaces::proto::Config::WithAsyncMethod_RunAction<akushon_interfaces::proto::Config::WithAsyncMethod_SetTorques<akushon_interfaces::proto::Config::WithAsyncMethod_SubscribeCurrentJoints<akushon_interfaces::proto::Config::Service> > > > > >*, grpc::ServerCompletionQueue*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
/usr/bin/ld: libakushon.so: undefined reference to `akushon::CallData<akushon_interfaces::proto::Empty, akushon_interfaces::proto::ConfigActions>::CallData(akushon_interfaces::proto::Config::WithAsyncMethod_GetConfig<akushon_interfaces::proto::Config::WithAsyncMethod_SaveConfig<akushon_interfaces::proto::Config::WithAsyncMethod_PublishSetJoints<akushon_interfaces::proto::Config::WithAsyncMethod_RunAction<akushon_interfaces::proto::Config::WithAsyncMethod_SetTorques<akushon_interfaces::proto::Config::WithAsyncMethod_SubscribeCurrentJoints<akushon_interfaces::proto::Config::Service> > > > > >*, grpc::ServerCompletionQueue*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)'
collect2: error: ld returned 1 exit status
`
I've actually already done that before mas at [this](https://github.com/ichiro-its/akushon/pull/37/commits/1c7559c2400d7f1de8c544d78924aa7425955fdc), but there's a request to put it all in the header file :pray:

Reply:
No need to move it to the `.cpp` file, just put it in the same header file under the class definition, like the example i showed.